src/gfx/main.cpp

```diff
@@ -840,18 +847,42 @@ auto Palette::begin() -> decltype(colors)::iterator {
 	// Skip the first slot if reserved for transparency
 	return colors.begin() + options.hasTransparentPixels;
 }
+
 auto Palette::end() -> decltype(colors)::iterator {
-	return std::find(begin(), colors.end(), UINT16_MAX);
+	// Since the palette may contain gaps, we must scan from the end.
+	for (auto it = colors.end(); ;) {
+		--it; // The array is not empty, so it's fine to decrement pre-emptively.
+		// If the iterator points at a non-empty element, then it is the end
+		// (being followed only by empty slots). Thus one past is is one past the end.
+		if (*it != UINT16_MAX) {
+			return ++it;
+		}
+		if (it == colors.begin()) { // Is it still safe to decrement?
+			return it; // No? Then the palette is empty.
+		}
+	}
 }
 
 auto Palette::begin() const -> decltype(colors)::const_iterator {
 	// Skip the first slot if reserved for transparency
 	return colors.begin() + options.hasTransparentPixels;
 }
+
 auto Palette::end() const -> decltype(colors)::const_iterator {
-	return std::find(begin(), colors.end(), UINT16_MAX);
+	// Since the palette may contain gaps, we must scan from the end.
+	for (auto it = colors.end(); ;) {
+		--it; // The array is not empty, so it's fine to decrement pre-emptively.
+		// If the iterator points at a non-empty element, then it is the end
+		// (being followed only by empty slots). Thus one past is is one past the end.
+		if (*it != UINT16_MAX) {
+			return ++it;
+		}
+		if (it == colors.begin()) { // Is it still safe to decrement?
+			return it; // No? Then the palette is empty.
+		}
+	}
```

Comment:
Whee, code duplication. :/ (Can we possibly have the `const` version delegate to the non-`const` one in C++?)
```suggestion
	// Since the palette may contain gaps, we must scan from the end.
	for (auto it = colors.end(); it != colors.begin(); --it) {
		// If the iterator points after a non-empty element, then it is the end
		// (being followed only by empty slots). Thus one past it is one past the end.
		if (*(it - 1) != UINT16_MAX) {
			return it;
		}
	}
	return colors.end(); // The palette is empty.
```
```

Reply:
You **can** delegate, by using... `const_cast`. It's okay as long as `this` isn't modified, but then you're "unbuckling the safety belt" so to speak.

The alternative is using macros. Not sure we wanna go *there*...

Reply:
Seems like casting non-`const` to `const` when it's returned would be safe and concise.

Reply:
Since the functions are close by and short, I'd rather live with that code duplicated.

Reply:
Okay, then this diff would need applying too to keep the `iterator` and `const_iterator` in sync.

Reply:
By the way, `*(it - 1)` can be written as `it[-1]`. It can even be a little faster for non-pointer iterators, since they can return the value instead of having to create a temporary iter :)